### DISCUSSION TITLE
could serverlessworkflow limit the parallelism of workflow instances?

### DISCUSSION THREAD

**Discussion by elim19** [serverlessworkflow/specification#750] [Q&A] [answered]
for example, the following state c will produce three subFlow simultaneously, could we limit  the parallelism of workflow subFlow?

![image](https://user-images.githubusercontent.com/123845880/215331393-a3c67168-f38c-4949-b257-d23b8b724216.png)

**Comment by tsurdilo** [comment]
with operation state you are execution actions sequentially in the order you are defining them in the array, you could use parallel state to execute them in parallel and then define how many you are awaiting to complete. Is this the question?

**Reply by elim19** [reply]
> ou are execution actions sequentially in the order you are defining them in the array, you could use parallel state to execute them in parallel and then define how many you are awaiting to complete. Is this the questio

@tsurdilo ，with parallel state,  all branches are executed simultaneously, can I limit the number of execution?
for example, the following execution has 3 branches, can I limit only 2 branches can be executed simultaneously,  and the last one has to wait until one of the first 2 branches is completed, then it can start?

<img width="519" alt="image" src="https://user-images.githubusercontent.com/123845880/215354916-75127dd4-cab2-48a0-a436-e1a62c7f4ff9.png">

**Comment by tsurdilo** [comment, marked as answer]
I see, so you have a batching use case, currently not possible with parallel state. 
best that we have currently is probably foreach state with mode set to "parallel" and define your batch size via the "batchSize" property, hope that helps, just note this still dont completely fit your particular use case which seems to be "sliding window" type batch approach

**Reply by tsurdilo** [reply]
i think we should add more support for batching use cases as they are widely used. things like batch size as well as sliding window  support would be useful as well. lets try to work on it

**Reply by tsurdilo** [reply]
also would look at the rate limiting extension here: https://github.com/serverlessworkflow/specification/blob/main/extensions/ratelimiting.md

specifically "maxConcurrentActions"
